Docker pull/push with max concurrency limits.

yongtang opened the pull request:
**\- What I did**

This fix tries to address issues raised in #20936 and #22443 where `docker pull` or `docker push` fails because of the concurrent connection failing.
Currently, the number of maximum concurrent connections is controlled by `maxDownloadConcurrency` and `maxUploadConcurrency` which are set to 3 and 5 respectively. Therefore, in situations where network connections don't support multiple downloads/uploads, failures may encounter for `docker push` or `docker pull`.

**\- How I did it**

This fix changes `maxDownloadConcurrency` and `maxUploadConcurrency` to adjustable by passing `--max-concurrent-uploads` and `--max-concurrent-downloads` to `docker daemon` command.

The documentation related to docker daemon has been updated.

**\- How to verify it**

Additional test case have been added to cover the changes in this fix.

**\- Description for the changelog**

Add  `--max--concurrent-uploads` and `--max--concurrent-downloads` to `docker daemon` command so that `docker pull` and `docker push` could control the max number of concurrent connections during uploads or downloads.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #20936. This fix fixes #22443.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

thaJeztah commented:
Thanks!

ping @aaronlehmann wdyt?

aaronlehmann commented:
Thanks for working on this.

I wasn't sure whether this should be a daemon-level flag or a flag that can be passed to the client and transmitted through the API. Now that I think about it some more, I think a client flag would be very unfriendly, because it would have to be specified on every pull. So I think the approach here makes sense.

I'd be interested to hear opinions on this, though.

aaronlehmann commented on daemon/config.go:
```diff
@@ -18,6 +18,11 @@ import (
 
 const (
+	defaultMaxDownloadConcurrency = 5
+	defaultMaxUploadConcurrency   = 3
```
I noticed that this reverses the old defaults. The old code was:

```
	// maxDownloadConcurrency is the maximum number of downloads that		
 	// may take place at a time for each pull.		
 	maxDownloadConcurrency = 3		
 	// maxUploadConcurrency is the maximum number of uploads that		
 	// may take place at a time for each push.		
 	maxUploadConcurrency = 5
```

I was a little surprised to see that there are more simultaneous uploads allowed than downloads, but I think I remember why it's like this. Registries often have high latencies to a storage backend, such as S3. At the end of a layer upload, there are multiple steps that happen to commit that upload, and because of the latency, this can take several seconds. If a push is uploading many small layers, it's faster overall to upload more at the same time, so less time is wasted in between uploads waiting for the registry to do these commits.

yongtang replied:
Thanks @aaronlehmann that was a typo I must have been confused about which one is which when I was working on it.

aaronlehmann commented on daemon/config.go:
```diff
@@ -138,6 +151,8 @@ func (config *Config) InstallCommonFlags(cmd *flag.FlagSet, usageFn func(string)
 	cmd.Var(opts.NewNamedMapOpts("cluster-store-opts", config.ClusterOpts, nil), []string{"-cluster-store-opt"}, usageFn("Set cluster store options"))
 	cmd.StringVar(&config.CorsHeaders, []string{"-api-cors-header"}, "", usageFn("Set CORS headers in the remote API"))
+	cmd.IntVar(&config.MaxDownloadConcurrency, []string{"-max-download-concurrency"}, defaultMaxDownloadConcurrency, usageFn("Set the max concurrent downloads for each pull"))
+	cmd.IntVar(&config.MaxUploadConcurrency, []string{"-max-upload-concurrency"}, defaultMaxUploadConcurrency, usageFn("Set the max concurrent uploads for each push"))
```
`-max-download-concurrency` is a long option, and concurrency is a relatively technical term.

What about `-simultaneous-downloads` / `-simultaneous-uploads`? @thaJeztah any ideas?

aaronlehmann commented on docs/reference/commandline/daemon.md:
```diff
@@ -898,6 +900,8 @@ This is a full example of the allowed configuration options in the file:
 	"cluster-store-opts": [],
 	"cluster-advertise": "",
+	"max-concurrency-download": 3,
+	"max-concurrency-upload": 5,
```
Shouldn't these be `max-download-concurrency` and `max-upload-concurrency`?

Do you think it would make sense to allow changing these settings at runtime? I think it would be nice to support that, especially since these are brand new options. It should just be a matter of adding a `SetConcurrency` method to `TransferManager`, and adding `SetConcurrency` methods to  `LayerUploadManager` and `LayerDownloadManager` which call `tm.SetConcurrency`.

MHBauer replied:
Yes, this doesn't match any of the other variable names.

Runtime config would be cool, but that obviously has to come from the client side.

aaronlehmann replied:
@MHBauer: I meant changing it at runtime by having the daemon reload the config file.

yongtang replied:
Thanks @aaronlehmann @MHBauer I just corrected the issue in the documentation. I also updated the code so that in case of reload options will be updated.

yongtang replied:
@aaronlehmann @MHBauer I temporarily leave the `--max-download/upload-concurrency` alone but let me know if we could find a short name so that I could update the PR.

LK4D4 commented:
I'm okay with flags. However now we need validation and probably a way to specify "unlimited".

MHBauer commented:
@aaronlehmann There is a client-side config file that could be used for this flag in the client.

I like 'max' something as the flag. That is the word I would first look for.

yongtang commented:
Thanks @LK4D4 I just updated the validation to the added flags. Please let me know if there are any issues.

aaronlehmann commented:
Thanks @yongtang, this looks good.

To address @LK4D4's suggestion of allowing unlimited concurrency, could we make `0` mean unlimited?

As for the flag names, maybe `-max-downloads` and `-max-uploads`?

thaJeztah commented:
Yes, was also thinking along the lines of `--max-downloads`, possibly `--max-concurrent-downloads`, but that's quite long again.

yongtang commented:
Thanks @LK4D4 @aaronlehmann @thaJeztah I just updated the pull request to change the names and make `0` as unlimited. Please let me know if there are any issues.

aaronlehmann commented:
LGTM

thaJeztah commented:
hm, before it's merged; should this setting be "reloadable"? https://docs.docker.com/engine/reference/commandline/daemon/#configuration-reloading, so that it can be re-configured without restarting the daemon?

aaronlehmann commented:
@thaJeztah: It already is; see the changes to the `Reload` function. I requested this during the review.

aaronlehmann commented:
Looks like this doc needs to be updated though - is that what you mean?

thaJeztah commented:
Oh! Missed that that was implemented now, didn't look at the code. But yes, it should be mentioned in the docs, below this header https://github.com/docker/docker/blob/master/docs/reference/commandline/dockerd.md#configuration-reloading

yongtang commented:
Thanks @aaronlehmann @thaJeztah I just updated the docs. Please let me know if there are any other issues.

thaJeztah commented on daemon/config.go:
```diff
@@ -378,5 +400,14 @@ func validateConfiguration(config *Config) error {
+	}
+
+	// validate MaxUploads
+	if config.IsValueSet("max-upload") && config.MaxUploads < 0 {
```
s/max-upload/max-uploads/ (plural)

yongtang replied:
Thanks a lot @thaJeztah. Sorry for missed that one typo before.

thaJeztah commented:
Ok I just tested this, and it looks like there's an issue; if I use a custom configuration, then remove that, or remove the "max-downloads" / "max-uploads", the values are not reset to their default:

```bash
echo "{\"max-uploads\":1}" > /etc/docker/daemon.json

# reload the daemon;
kill -1 13

# INFO[0847] Got signal to reload configuration, reloading from: /etc/docker/daemon.json
# DEBU[0847] Reset Max Concurrent Uploads: 1

rm  /etc/docker/daemon.json

# reload the daemon;
kill -1 13

# INFO[0922] Got signal to reload configuration, reloading from: /etc/docker/daemon.json
# ERRO[0922] open /etc/docker/daemon.json: no such file or directory

# same result (configuration file present, but "max-downloads" not set):
echo "{\"labels\":[\"foo=bar\"]}" > /etc/docker/daemon.json

# reload the daemon;
kill -1 13

# INFO[1007] Got signal to reload configuration, reloading from: /etc/docker/daemon.json
# (max-downloads is not reset)
```

thaJeztah commented on daemon/daemon.go:
```diff
@@ -1505,6 +1500,36 @@ func (daemon *Daemon) Reload(config *Config) error {
 	}
+	// If no value is set for max-downloads
+	// we assume it is the default value
+	maxDownloads := defaultMaxDownloads
```
Guess the problem is this; if it's no longer set, `maxUploads` is set to the default, so  
`maxDownloads != originalMaxDownloads` is never true.

yongtang commented:
Thanks @thaJeztah I take a look and the issue is caused by the fact that, when new `config` is passed to update the daemon configurations (`configStore`), the `IsValueSet` is not updated in `daemon.configStore`.
(The `IsValueSet` is not updated for all reload configurations including `label` and `debug`).

For `MaxUploads/MaxDownloads`, since we are just updating the concurrencyLimit which is not a heavyweight operation, I think it might be more straightforward and easier to maintain by simply `always set the value (with empty field for default value)`. This will also allows user to reset the default value by leave the field of `MaxUploads/MaxDownloads` empty.

I updated the PR based on the above assumption but please let me know if we prefer other ways to handle the mentioned scenario.

yongtang commented:
@thaJeztah On a side note, I noticed that the labels and debug fields are updated through the following:
```
	if config.IsValueSet("labels") {
		daemon.configStore.Labels = config.Labels
	}
	if config.IsValueSet("debug") {
		daemon.configStore.Debug = config.Debug
	}
```

That means if user skip the debug field, daemon will assume the expected behavior to be `continue with current mode` (could be `debug=true` or `debug=false` depending on the previous state).
Not sure if this could be a concern?
Since for the above case user will need to check the daemon information to know what is the expected result (instead of knowing what will happen when submitting a config file).

thaJeztah commented:
@yongtang yes, this looks better, one (possibly edge case); `null` is currently converted to `0`, so not specifying a value sets the max downloads to `unlimited`.  So this leads to setting "unlimited" downloads;

```json
{"max-downloads":null}
```

I'm not sure what the expected outcome should be there, and possibly we cannot prevent that (if it's unmarshalled to an integer. Ideally, I'd say "null" equals to "not set", so set it to the defaults.


w.r.t. "labels" and "debug"; I agree, not setting a value should reset it to the default; perhaps you can open a new PR for that?

Thanks so much for your continuing contributions, @yongtang, good to have you around!

yongtang commented:
@thaJeztah Thanks for your thorough review and help. Much appreciated. Let me take a look at the `null` issue and see what we could do with it. I also opened another PR #22532 to cover the "labels" and "debug" issue. Again thanks for the help!

yongtang committed:
Docker pull/push with max concurrency limits.

This fix tries to address issues raised in #20936 and #22443
where `docker pull` or `docker push` fails because of the
concurrent connection failing.
Currently, the number of maximum concurrent connections is
controlled by `maxDownloadConcurrency` and `maxUploadConcurrency`
which are hardcoded to 3 and 5 respectively. Therefore, in
situations where network connections don't support multiple
downloads/uploads, failures may encounter for `docker push`
or `docker pull`.

This fix tries changes `maxDownloadConcurrency` and
`maxUploadConcurrency` to adjustable by passing
`--max-concurrent-uploads` and `--max-concurrent-downloads` to
`docker daemon` command.

The documentation related to docker daemon has been updated.

Additional test case have been added to cover the changes in this fix.

This fix fixes #20936. This fix fixes #22443.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@thaJeztah I updated the PR so that config file will be sanitized by unmarshal/marshal with null value fields removed. This will resolve the issue of situations like `{"max-downloads":null}`. Please let me know if there are any issues.

thaJeztah commented:
Thanks @yongtang, LGTM

ping @aaronlehmann PTAL

aaronlehmann commented on daemon/config.go:
```diff
@@ -218,6 +239,28 @@ func getConflictFreeConfiguration(configFile string, flags *flag.FlagSet) (*Conf
 
+	// Sanitize values of null so that configurations like
+	// `{ "max-downloads" : null }` will not be considered as 0.
+	// Note: Only the first level is checked.
```
I think a better way to do this is to make `MaxDownloads` and `MaxUploads` `*int` instead of `int`. They will be nil pointers unless the value can be unmarshalled as an integer.

yongtang commented:
Thanks @aaronlehmann. I changed the `MaxDownloads` and `MaxUploads` from `int` to `*int`. Let me know if there are any other issues.

aaronlehmann commented on daemon/config.go:
```diff
@@ -218,6 +244,28 @@ func getConflictFreeConfiguration(configFile string, flags *flag.FlagSet) (*Conf
+		if jsonBytes, err := json.Marshal(jsonData); err == nil {
+			b = jsonBytes
+		}
+	}
```
You shouldn't need this extra `Unmarshal`/`Marshal`. It should be enough to check if `MaxDownloads` or `MaxUploads` is `nil` after unmarshalling, now that they are pointers.

yongtang replied:
@aaronlehmann Sorry I was meant to remove this chunk of code in my update. Somehow it was not deleted with git. Let me update the pull request again.

aaronlehmann commented on daemon/daemon.go:
```diff
@@ -1520,6 +1515,33 @@ func (daemon *Daemon) Reload(config *Config) error {
+	// If no value is set for max-downloads we assume it is the default value
+	// We always "reset" as the cost is lightweight and easy to maintain.
+	if config.IsValueSet("max-downloads") && config.MaxDownloads != nil {
+		daemon.configStore.MaxDownloads = config.MaxDownloads
```
This is probably fine in practice, but I think it would be slightly better to do:

```go
*daemon.configStore.MaxDownloads = *config.MaxDownloads
```

so that if something changes `config.MaxDownloads` after calling `Reload`, it doesn't unexpectedly change `daemon.configStore.MaxDownloads`.

yongtang replied:
@aaronlehmann Thanks just updated the pull request.

aaronlehmann commented on daemon/daemon.go:
```diff
@@ -719,8 +710,10 @@ func NewDaemon(config *Config, registryService *registry.Service, containerdRemo
 
-	d.downloadManager = xfer.NewLayerDownloadManager(d.layerStore, maxDownloadConcurrency)
-	d.uploadManager = xfer.NewLayerUploadManager(maxUploadConcurrency)
+	logrus.Debugf("Max Concurrent Downloads: %d", *config.MaxDownloads)
```
I notice that the pointers aren't checked for nil in this function. I *think* that's okay, because when this function is called, they should already have been initialized by `InstallCommonFlags`. Is that right?

yongtang replied:
@aaronlehmann Yes `config.MaxDownloads/MaxUploads` should have been initialized by `InstallCommonFlags`.

aaronlehmann commented:
LGTM

vdemeester commented:
@thaJeztah documentation updates right ?
@yongtang `man/dockerd.8.md` should be updated too :angel:

thaJeztah commented:
docs look good to me, but it needs to be added to the manage indeed. I'm thinking of opening a follow up PR to add some info to `docker pull` and `docker push` after this is merged

LK4D4 commented:
@thaJeztah so, maybe we can merge it and proceed with pure docs PR?

yongtang commented:
@aaronlehmann @vdemeester @LK4D4 @thaJeztah Thanks for the help. The man docs have been updated. Let me know if there are any other issues.

thaJeztah commented:
@LK4D4 I think we can wait one day so that @yongtang can add the man page (the "primary" documentation is complete then for this feature)

thaJeztah commented:
Haha, so no day waiting is needed. LGTM

icecrime commented:
Sorry but I really dislike the names `--max-uploads` and `--max-downloads`, it feels like I'm going to have to pay after the first n images downloaded :confused:

Explicit is better than short, can we please rename to `--max-concurrent-downloads` and `--max-concurrent-uploads`? Thanks, and sorry for being late to the party.

yongtang commented:
Thanks @icecrime just updated the PR. Please let me know if there are any other issues.

aaronlehmann commented:
re-LGTM

thaJeztah commented:
LGTM (`--max-concurrent-downloads` was my second choice, so 👍 )

icecrime commented:
@yongtang Thanks a lot for the rapid update!
@aaronlehmann @thaJeztah Thanks a lot for the rapid re-review!

thaJeztah commented:
ping @albers @sdurrheimer think this needs changes to the completion scripts. Sorry forgot to ping earlier :heart:

albers commented:
I'll add it to bash completion. Thanks for the ping.
